validate context and lines removed by the patch

wfraser opened the pull request:
This change validates that the lines of context and the lines removed actually match the text. If they don't, cargo-patch will abort with an error containing the filename and line number that failed to match.

Fixes #3

mettke commented:
Very nice addition thanks. I, of course, would love if you find the time to write a small test here for the failure case:
https://github.com/itmettkeDE/cargo-patch/blob/master/tests/patch_crates_io.rs

This way we can make it a bit more unlikely that future changes accidentally change expected behavior. Also it seems like you have to run `cargo fmt` 🙂

wfraser commented:
Yeah, good point. I've added a couple tests and ran rustfmt.

Note that a couple of the existing tests had to be changed:
 - one had a patch with an extra empty line at the end, which wouldn't pass because that's considered mismatched context; I assume that wasn't intentional so I just removed that line
 - another one intentionally had mismatched context and apparently was intended to succeed? That's now considered an error, so I changed the test to check for failure instead.

Also note that as a bit of a drive-by change, I modified the success output to include the file being patched. Currently on success it just prints `Patched <crate name>` for each file, which looks a little strange if you're patching multiple files in a crate.

mettke commented:
Very nice - looks good. But it seems like there are two more lines in the test where the formatter is complaining 🤔:

```rust
Diff in /home/runner/work/cargo-patch/cargo-patch/src/main.rs at line 394:
 This is the third line
 "#;
         let patch = Patch::from_single(patch).expect("Unable to parse patch");
-        let test_patched = apply_patch(patch, content).expect("Failed to apply patch");
+        let test_patched =
+            apply_patch(patch, content).expect("Failed to apply patch");
         assert_eq!(patched, test_patched, "Patched content does not match");
     }
 
Diff in /home/runner/work/cargo-patch/cargo-patch/src/main.rs at line 43[9](https://github.com/itmettkeDE/cargo-patch/runs/8294878976?check_suite_focus=true#step:3:10):
 id est laborum.
 "#;
         let patch = Patch::from_single(patch).expect("Unable to parse patch");
-        let test_patched = apply_patch(patch, content).expect("Failed to apply patch");
+        let test_patched =
+            apply_patch(patch, content).expect("Failed to apply patch");
         assert_eq!(patched, test_patched, "Patched content does not match");
     }
```

wfraser commented:
Sorry about that. Fixed.

mettke commented:
No worries. Thanks for the contribution! Will be released soon 🙂
